### PR TITLE
Fix broken regressions

### PR DESCRIPTION
**Pull request by waynemitchell**
Fixing the regression tests broken by https://github.com/hypre-space/hypre/pull/1174

### CODE REVIEW

File: src/struct_mv/box.c [hunk 1]
```diff
@@ -1095,12 +1095,14 @@ hypre_BoxArrayCreateFromIndices( HYPRE_Int         ndim,
             }
 
             /* Look for largest sign change among all directions */
-            dir = -1;
+            dir = 0;
```
Comment (rfalgout) [inline src/struct_mv/box.c:1098]: Is the issue that `dir` needs to be a valid direction (set to `0` by default) even when no sign change is detected?

Comment (waynemitchell) [reply src/struct_mv/box.c:1098]: Yeah, you can see that `dir` is used in line 1102 below. I was trying to just reuse this variable for convenience, but I wasn't careful enough about it.

This should fix the valgrind issues in the regressions, but there were also some iteration count differences reported in some error files. Some of these are expected: namely any place that the unstructured interpolation for SSAMG is used, the change here may change some of the numerics. But I think there are still some tests where iteration counts change that I don't quite understand.

**Comment by waynemitchell** [comment]
OK, I found the culprit for the regression changes in the amr2d and amr3d regressions (and a general reason why behavior will change here). The `hypre_BoxArrayCreateFromIndices()` function is used here in assembling the sstruct graph:
https://github.com/hypre-space/hypre/blob/recmat/src/sstruct_mv/HYPRE_sstruct_graph.c#L763
My initial change affects the behavior of how the boxes are generated where there is no sign change in the Laplacian when running the Berger-Rigoutsos algorithm. I don't know this algorithm well enough to understand exactly when this happens... But I do see that this occurs somewhat frequently. The previous behavior of the `hypre_BoxArrayCreateFromIndices()` code was to split the box in dim 0 by default. This can cause issues if the box already has no thickness in dim 0. My change avoids this issue by splitting along the dimension where the box is largest. Again, I don't know the underlying algorithm well here, so I'm not sure if there are other implications here that I'm not aware of, but I think my change is at least an improvement in that it prevents the code from trying to split a box along dim 0 when it cannot (which causes it to error out). If there are no additional concerns from @victorapm or @rfalgout , I think we should go ahead and merge this fix and then updated the .saved files to account for the changes in convergence due to the change in behavior here.

**Comment by victorapm** [comment]
﻿Thanks for the update, Wayne!

Your idea sounds good to me, but I think we should return to this at some point since convergence has degraded for a few cases with the new changes

Victor

From: Wayne Mitchell ***@***.***>
Date: Friday, November 8, 2024 at 1:34 PM
To: hypre-space/hypre ***@***.***>
Cc: Paludetto Magri, Victor ***@***.***>, Mention ***@***.***>
Subject: Re: [hypre-space/hypre] Fix broken regressions (PR #1180)

OK, I found the culprit for the regression changes in the amr2d and amr3d regressions (and a general reason why behavior will change here). The hypre_BoxArrayCreateFromIndices() function is used here in assembling the sstruct graph:
https://github.com/hypre-space/hypre/blob/recmat/src/sstruct_mv/HYPRE_sstruct_graph.c#L763<https://urldefense.us/v3/__https:/github.com/hypre-space/hypre/blob/recmat/src/sstruct_mv/HYPRE_sstruct_graph.c*L763__;Iw!!G2kpM7uM-TzIFchu!1sPoyOHjaC3kunl1LWB2EBW-PpoqQNb8-gdZXyl4vzS3LfiZzJ67Hy_b23DD3AzgjsMWQcuNwBdv1m-epQddEDZ_jLs_gRc$>
My initial change affects the behavior of how the boxes are generated where there is no sign change in the Laplacian when running the Berger-Rigoutsos algorithm. I don't know this algorithm well enough to understand exactly when this happens... But I do see that this occurs somewhat frequently. The previous behavior of the hypre_BoxArrayCreateFromIndices() code was to split the box in dim 0 by default. This can cause issues if the box already has no thickness in dim 0. My change avoids this issue by splitting along the dimension where the box is largest. Again, I don't know the underlying algorithm well here, so I'm not sure if there are other implications here that I'm not aware of, but I think my change is at least an improvement in that it prevents the code from trying to split a box along dim 0 when it cannot (which causes it to error out). If there are no additional concerns from @victorapm<https://urldefense.us/v3/__https:/github.com/victorapm__;!!G2kpM7uM-TzIFchu!1sPoyOHjaC3kunl1LWB2EBW-PpoqQNb8-gdZXyl4vzS3LfiZzJ67Hy_b23DD3AzgjsMWQcuNwBdv1m-epQddEDZ_aS53o7w$> or @rfalgout<https://urldefense.us/v3/__https:/github.com/rfalgout__;!!G2kpM7uM-TzIFchu!1sPoyOHjaC3kunl1LWB2EBW-PpoqQNb8-gdZXyl4vzS3LfiZzJ67Hy_b23DD3AzgjsMWQcuNwBdv1m-epQddEDZ_B6YfpZM$> , I think we should go ahead and merge this fix and then updated the .saved files to account for the changes in convergence due to the change in behavior here.

**Comment by rfalgout** [comment]
> I think we should return to this at some point since convergence has degraded for a few cases with the new changes

Which regression tests are affected?  Do you need one of use to generate the saved file changes on a tux box?  I would like to see what the impact is, whether we address it now or later.

**Comment by waynemitchell** [comment]
> > I think we should return to this at some point since convergence has degraded for a few cases with the new changes
> 
> Which regression tests are affected? Do you need one of use to generate the saved file changes on a tux box? I would like to see what the impact is, whether we address it now or later.

@rfalgout , yes, can you generate updated .saved files (since I'm still on RHEL 8 and may not get exactly the right results on mine)?

**Comment by rfalgout** [comment]
Okay, the saved files are there now.

**Review by victorapm — APPROVED** [review]

**Comment by rfalgout** [comment]
@waynemitchell  Do you want me to run the tux tests on this?  Let me know if you think it's ready.  Thanks!

**Review by rfalgout — APPROVED** [review]
I agree with @victorapm that we should look into the convergence changes at some point later.  This looks good otherwise.  Thanks!

**Comment by rfalgout** [comment]
The sstruct regression tests on my tux box are clean.  I think this is ready to merge (it's up to date with master too).  Thanks!

**Pull request merged**